policystore/ipset_test.go

```diff
@@ -0,0 +1,157 @@
+
+	uut := NewIPSet(proto.IPSetUpdate_IP)
+	addr := envoyapi.Address{Address: &envoyapi.Address_SocketAddress{
+		SocketAddress: &envoyapi.SocketAddress{
+			Address: "2.2.2.2",
+		},
+	}}
+
+	uut.RemoveString("2.2.2.2")
+	g.Expect(uut.ContainsAddress(&addr)).To(BeFalse())
+	uut.AddString("2.2.2.2")
+	g.Expect(uut.ContainsAddress(&addr)).To(BeTrue())
+	uut.RemoveString("2.2.2.2")
+	g.Expect(uut.ContainsAddress(&addr)).To(BeFalse())
+	// Test idempotency
+	uut.RemoveString("2.2.2.2")
+	g.Expect(uut.ContainsAddress(&addr)).To(BeFalse())
+
+	// Adding a different address should not affect a removed one
+	uut.AddString("2.2.2.3")
+	g.Expect(uut.ContainsAddress(&addr)).To(BeFalse())
+}
+
+func TestAddIpAndPort(t *testing.T) {
+	g := NewGomegaWithT(t)
+
+	uut := NewIPSet(proto.IPSetUpdate_IP_AND_PORT)
+	addr := envoyapi.Address{Address: &envoyapi.Address_SocketAddress{
+		SocketAddress: &envoyapi.SocketAddress{
+			Address:       "2.2.2.2",
+			Protocol:      envoyapi.SocketAddress_TCP,
+			PortSpecifier: &envoyapi.SocketAddress_PortValue{PortValue: 2222},
+		},
+	}}
+
+	uut.AddString("2.2.2.2,tcp:2222")
+	g.Expect(uut.ContainsAddress(&addr)).To(BeTrue())
+
+	// Different port.
+	addr.GetSocketAddress().GetPortSpecifier().(*envoyapi.SocketAddress_PortValue).PortValue = 33
+	g.Expect(uut.ContainsAddress(&addr)).To(BeFalse())
+
+	// Add other port
+	uut.AddString("2.2.2.2,tcp:33")
+	g.Expect(uut.ContainsAddress(&addr)).To(BeTrue())
+}
+
+func TestRemoveIpAndPort(t *testing.T) {
+	g := NewGomegaWithT(t)
+
+	uut := NewIPSet(proto.IPSetUpdate_IP_AND_PORT)
+	addr := envoyapi.Address{Address: &envoyapi.Address_SocketAddress{
+		SocketAddress: &envoyapi.SocketAddress{
+			Address:       "2.2.2.2",
+			Protocol:      envoyapi.SocketAddress_TCP,
+			PortSpecifier: &envoyapi.SocketAddress_PortValue{PortValue: 2222},
+		},
+	}}
+
+	uut.RemoveString("2.2.2.2,tcp:2222")
+	g.Expect(uut.ContainsAddress(&addr)).To(BeFalse())
+	uut.AddString("2.2.2.2,tcp:2222")
+	g.Expect(uut.ContainsAddress(&addr)).To(BeTrue())
+	uut.RemoveString("2.2.2.2,tcp:2222")
+	g.Expect(uut.ContainsAddress(&addr)).To(BeFalse())
+	// Test idempotency
+	uut.RemoveString("2.2.2.2,tcp:2222")
+	g.Expect(uut.ContainsAddress(&addr)).To(BeFalse())
+
+	// Adding a different address should not affect a removed one
+	uut.AddString("2.2.2.3,tcp:2222")
+	g.Expect(uut.ContainsAddress(&addr)).To(BeFalse())
+}
+
+func TestIpPortContainsAddress(t *testing.T) {
+	g := NewGomegaWithT(t)
+
+	uut := NewIPSet(proto.IPSetUpdate_IP_AND_PORT)
+	addr := envoyapi.Address{Address: &envoyapi.Address_SocketAddress{
+		SocketAddress: &envoyapi.SocketAddress{
+			Address:       "2.2.2.2",
+			Protocol:      envoyapi.SocketAddress_TCP,
+			PortSpecifier: &envoyapi.SocketAddress_PortValue{PortValue: 2222},
+		},
+	}}
+
+	// Different protocol
+	uut.AddString("2.2.2.2,udp:2222")
+	g.Expect(uut.ContainsAddress(&addr)).To(BeFalse())
+
+	sa := addr.GetSocketAddress()
+	sa.Protocol = envoyapi.SocketAddress_UDP
```

Comment:
nit: Could we create two separate instances of Address one with protocol UDP and another with TCP. It would help with readability of the test.
same suggestion for ports and ip's. it's your call.

Reply:
I did it that way because the `Addr` proto initialization is long winded. So, I added a `makeAddr` function to make it more manageable.